File: src/main/java/com/lessmarkup/interfaces/recordmodel/RecordModelDefinition.java
package com.lessmarkup.interfaces.recordmodel;

import com.google.gson.JsonElement;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.exceptions.RecordValidationException;
import java.util.List;

public interface RecordModelDefinition {
    String getTitleTextId();
    String getModuleType();
    Class<? extends RecordModel> getModelType();
    Class<? extends DataObject> getDataType();
    String getId();
    List<InputFieldDefinition> getFields();
    List<RecordColumnDefinition> getColumns();
    void validateInput(JsonElement objectToValidate, boolean isNew) throws RecordValidationException;
    boolean isSubmitWithCaptcha();
    ModelCollection createModelCollection();
}

File: src/main/scala/com/lessmarkup/interfaces/recordmodel/RecordModelDefinition.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.recordmodel

import com.google.gson.JsonElement
import com.lessmarkup.interfaces.data.DataObject

trait RecordModelDefinition {
  def getTitleTextId: String

  def getModuleType: String

  def getModelType: Class[_ <: RecordModel[_]]

  def getDataType: Class[_ <: DataObject]

  def getId: String

  def getFields: List[InputFieldDefinition]

  def getColumns: List[RecordColumnDefinition]

  def validateInput(objectToValidate: JsonElement, isNew: Boolean)

  def isSubmitWithCaptcha: Boolean

  def createModelCollection: ModelCollection[_]
}